File: perceptron/perceptron.R
### 初期値を設定
test_data <- matrix(c(1, 1, 1, 1, -1.3, -0.2, 0.5, 1), 4, 2) # 学習データを行列で用意
label <- c(-1, -1, 1, 1) # 学習データのラベル
w <- c(0.2, 0.3) # 初期重み
rho <- 0.5 # 学習係数

### 学習データをプロット
plot(cbind(test_data[1:2, 2], rep(0, 2)), col="red",
     xlim=c(-2, 2), ylim=c(-2,2), xlab="x", ylab="y")
par(new=T)
plot(cbind(test_data[3:4, 2], rep(0, 2)), col="blue",
     xlim=c(-2, 2), ylim=c(-2,2), xlab="x", ylab="y")

### 重みベクトルを更新する関数
update <- function(x, l, w) {
  if (sign(x %*% w) == sign(l)) {
    return (w)
  }
  return (w+rho*x*l)
}

### 全データで重みベクトルが更新されなくなるまで繰り返す
count <- 0 
while (count < nrow(test_data)) {
  for (i in 1:nrow(test_data)) {
    temp <- update(test_data[i,], label[i], w)
    if (all(temp == w)) {
      count <- count+1
    } else {
      cat("クラスを誤認識しました\n")
      cat("重みベクトルを更新します\n")
      w <- temp
      print(w)
      count <- 0
    }
  }
}

### 最終的な識別関数を図示
curve(w[2]*x+w[1], -5, 5, add=T)

File: perceptron/perceptron.py
# coding: UTF-8
import numpy as np
import math
import sys

### 符号を調べる関数
# Rでいうsign関数が見当たらないので、とりあえず実装
def sign(x):
	if x > 0:
		return 1
	else:
		return -1

### 初期値を設定
test_data = np.array([[1, -1.3], [1, -0.2], [1, 0.5], [1, 1]])
label = np.array([-1, -1, 1, 1])
w = np.array([0.2, 0.3])
rho = 0.5

### 重みベクトルを更新する関数
def update(x, l, w):
	if (sign(np.dot(x, w)) == sign(l)):
		return w
	return w + rho * x * l

### 全データで重みベクトルが更新されなくなるまで繰り返す
count = 0 
while (count < test_data.shape[0]):
	for i in range(0, test_data.shape[0]):
		temp = update(test_data[i], label[i], w)
		if (all(temp == w)):
			count = count+1
		else:
			print("クラスを誤認識しました\n")
			print("重みベクトルを更新します\n")
			w = temp
			print(w)
			count = 0

print("識別関数は、%f x %fです" % (w[1], w[0]))